pandas/core/frame.py

```diff
@@ -6892,56 +6892,93 @@ def cov(self, min_periods=None):
 
         return self._constructor(baseCov, index=idx, columns=cols)
 
-    def corrwith(self, other, axis=0, drop=False):
+    def corrwith(self, other, axis=0, how='pairwise', drop=False,
+                 method='pearson', min_periods=1):
         """
-        Compute pairwise correlation between rows or columns of two DataFrame
-        objects.
+        Compute correlation between rows or columns of two DataFrame objects
+        or DataFrame and Series objects.
 
         Parameters
         ----------
         other : DataFrame, Series
         axis : {0 or 'index', 1 or 'columns'}, default 0
             0 or 'index' to compute column-wise, 1 or 'columns' for row-wise
+        how : {'pairwise', 'all'}, default 'pairwise' if other is a DataFrame
+            and default 'all' if other is a Series object
+            * 'pairwise' to compute correlation between rows or columns with
+              the same name/index,
+            * 'all' to compute correlation between all possible pairs of rows
+              or columns
         drop : boolean, default False
             Drop missing indices from result, default returns union of all
+        method : {'pearson', 'kendall', 'spearman'}
+            * pearson : standard correlation coefficient
+            * kendall : Kendall Tau correlation coefficient
+            * spearman : Spearman rank correlation
+        min_periods : int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result. Currently only available for pearson
+            and spearman correlation
+
+        Returns
+        -------
+        y : DataFrame
 
         Returns
         -------
-        correls : Series
+        correl : Series (how='pairwise') or DataFrame (how='all')
         """
+
         axis = self._get_axis_number(axis)
         this = self._get_numeric_data()
 
         if isinstance(other, Series):
-            return this.apply(other.corr, axis=axis)
+            return this.apply(other.corr, axis=axis,
+                              method=method, min_periods=min_periods)
 
-        other = other._get_numeric_data()
+        if not isinstance(other, DataFrame):
+            raise TypeError("'other' parameter should be a DataFrame "
+                            "or a Series object")
 
-        left, right = this.align(other, join='inner', copy=False)
```

Comment:
its odd that things didn't break when you removed this, have you tested a left / right that don't align? (IOW different columns / rows)

Reply:
They get aligned further in `pd.Series.corr()`:
```python
this, other = self.align(other, join='inner', copy=False)
```